### PR TITLE
STORM-3528: Allow users to provide their own custom TriggerPolicy/EvictionPolicy in BaseWindowBolt

### PR DESCRIPTION
**Pull request by joshi-aparna**

### CODE REVIEW

**Comment by srdo** [comment]
Please correct the style violations, you can see them at the bottom of the build here https://travis-ci.org/apache/storm/jobs/601641538?utm_medium=notification&utm_source=github_status

Running the build locally will also print them to console, using `mvn validate`

**Comment by joshi-aparna** [comment]
@srdo thank you. The checks are passing now.

File: storm-client/src/jvm/org/apache/storm/topology/WindowedBoltExecutor.java [hunk 1]
```diff
@@ -205,10 +205,21 @@ protected void validate(Map<String, Object> topoConf, Count windowLengthCount, D
         }
         // validate
         validate(topoConf, windowLengthCount, windowLengthDuration,
-                 slidingIntervalCount, slidingIntervalDuration);
-        evictionPolicy = getEvictionPolicy(windowLengthCount, windowLengthDuration);
-        triggerPolicy = getTriggerPolicy(slidingIntervalCount, slidingIntervalDuration,
-                                         manager, evictionPolicy);
+            slidingIntervalCount, slidingIntervalDuration);
+        if (topoConf.containsKey(Config.TOPOLOGY_BOLTS_WINDOW_EVICTION_POLICY)
+            && topoConf.get(Config.TOPOLOGY_BOLTS_WINDOW_EVICTION_POLICY) instanceof EvictionPolicy) {
```
Comment (srdo) [inline storm-client/src/jvm/org/apache/storm/topology/WindowedBoltExecutor.java]: I don't think you need to do the instanceof check here, we already specify that it needs to be that type in Config.

File: storm-client/src/jvm/org/apache/storm/topology/WindowedBoltExecutor.java [hunk 2]
```diff
@@ -205,10 +205,21 @@ protected void validate(Map<String, Object> topoConf, Count windowLengthCount, D
...
+            evictionPolicy = (EvictionPolicy<Tuple, ?>)topoConf.get(Config.TOPOLOGY_BOLTS_WINDOW_EVICTION_POLICY);
+
+        } else {
+            evictionPolicy = getEvictionPolicy(windowLengthCount, windowLengthDuration);
+        }
+        if (topoConf.containsKey(Config.TOPOLOGY_BOLTS_WINDOW_TRIGGER_POLICY)
+            && topoConf.get(Config.TOPOLOGY_BOLTS_WINDOW_TRIGGER_POLICY) instanceof TriggerPolicy) {
```
Comment (srdo) [inline storm-client/src/jvm/org/apache/storm/topology/WindowedBoltExecutor.java]: Also here

File: storm-client/src/jvm/org/apache/storm/Config.java [hunk 3]
```diff
@@ -1676,6 +1678,18 @@
     //DO NOT CHANGE UNLESS WE ADD IN STATE NOT STORED IN THE PARENT CLASS
     private static final long serialVersionUID = -1550278723792864455L;
 
+    /**
+     * Specify the trigger policy for the window.
+     */
+    @IsType(type = TriggerPolicy.class)
+    public static final String TOPOLOGY_BOLTS_WINDOW_TRIGGER_POLICY = "topology.bolts.window.trigger.policy";
```
Comment (srdo) [inline storm-client/src/jvm/org/apache/storm/Config.java]: Please move these to be next to the other window-related properties, such as sliding count.

**Comment by kishorvpatil** [comment]
@joshi-aparna Could you please update the PR?

**Comment by joshi-aparna** [comment]
> @joshi-aparna Could you please update the PR?

Done. Merged from master

**Pull request closed without merging**